--- nmap-2.53/tcpip.h.orig	Tue May 16 18:21:58 2000
+++ nmap-2.53/tcpip.h	Tue May 16 18:22:09 2000
@@ -354,9 +354,9 @@
 /* Hex dump */
 int get_link_offset(char *device);
 char *readip_pcap(pcap_t *pd, unsigned int *len, long to_usec);
-#ifndef HAVE_INET_ATON
+/*#ifndef HAVE_INET_ATON
 int inet_aton(register const char *, struct in_addr *);
-#endif
+#endif*/
 
 
 #endif /*TCPIP_H*/
--- nmap-2.53/nmap.h.orig Tue May 16 18:18:00 2000
+++ nmap-2.53/nmap.h  Tue May 16 18:18:30 2000
@@ -362,9 +362,9 @@
 void skid_output(unsigned char *s);

 /* From glibc 2.0.6 because Solaris doesn't seem to have this function */
-#ifndef HAVE_INET_ATON
+/*#ifndef HAVE_INET_ATON
 int inet_aton(register const char *, struct in_addr *);
-#endif
+#endif*/
 #ifndef HAVE_SNPRINTF
 int snprintf ( char *str, size_t n, const char *format, ... );
 #endif
